src/lua/datetime.lua

```diff
@@ -172,11 +293,96 @@ local function check_ymd_range(y, M, d)
         return
     end
 ::min_err::
-    error(('date %d-%02d-%02d is less than minimum allowed %d-%02d-%02d'):
-        format(y, M, d, MIN_DATE_YEAR, MIN_DATE_MONTH, MIN_DATE_DAY))
+    error(('date %d-%02d-%02d is less than minimum allowed %s'):
+        format(y, M, d, MIN_DATE_TEXT))
 ::max_err::
-    error(('date %d-%02d-%02d is greater than maximum allowed %d-%02d-%02d'):
-        format(y, M, d, MAX_DATE_YEAR, MAX_DATE_MONTH, MAX_DATE_DAY))
+    error(('date %d-%02d-%02d is greater than maximum allowed %s'):
+        format(y, M, d, MAX_DATE_TEXT))
+end
+
+-- check v value against maximum/minimum possible values
+-- if there is nil then simply return default 0
+local function checked_max_value(v, max, txt)
+    if v == nil then
+        return 0
+    end
+    if type(v) ~= 'number' then
+        error(('numeric value expected, but received %s'):
+              format(type(v)), 2)
+    end
+    if v > -max and v < max then
+        return v
+    end
+    error(('value %s of %s is out of allowed range [%s, %s]'):
+            format(v, txt, -max, max), 4)
+end
+
+
+local function bool2int(b)
+    return b and 1 or 0
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 or nsec >= NANOS_PER_SEC then
+        secs = secs + math_floor(nsec / NANOS_PER_SEC)
+        nsec = nsec % NANOS_PER_SEC
+    end
+    return secs, nsec
```

Comment:
Thanks for the report - that was good one!

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -627,8 +627,21 @@ local function qtail(s)
     return #s ~= 0 and (s .. ', ') or s
 end
 
+-- if sec is negative number, and nsec normalized then we denormalize sec
+-- and move toward 0
+local function normalize_neg_nsec(sec, nsec)
+    if sec < 0 and nsec ~= 0 then
+        assert(nsec > 0)
+        nsec = NANOS_PER_SEC - nsec
+        sec = sec + 1
+    end
+    return sec, nsec
+end
+
+-- signed or unsigned textual representation of sec and nsec
 local function seconds_str(need_sign, sec, nsec)
     local prefix = need_sign and '%+d' or '%d'
+    sec, nsec = normalize_neg_nsec(sec, nsec)
     sec = math_fmod(sec, 60)
     return nsec ~= 0 and (prefix..'.%09d seconds'):format(sec, nsec) or
            (prefix..' seconds'):format(sec)
```

which changes behavior here:
```
tarantool> inew = date.interval.new
---
...

tarantool> i1 = inew{nsec = 999999999} i2 = inew{nsec = 2}
---
...

tarantool> i1 + i2
---
- +1.000000001 seconds
...

tarantool> i1 = inew{nsec = -999999999} i2 = inew{nsec = -2}
---
...

tarantool> i1 + i2
---
- -1.000000001 seconds
...
```

Reply:
They are still broken:
```
interval = require('datetime').interval
i1 = interval.new({nsec = -999999999})
i2 = interval.new({nsec = -2})
i = i1 + i2
tarantool> i.sec, i.nsec
---
- -2
- 999999999
...
-- Garbage
tarantool> interval.new({nsec = -3000000000})
---
- -2.147483648 seconds
...
-- Wrong sign.
tarantool> interval.new({nsec = -300000000})
---
- +0.300000000 seconds
...
-- Garbage but via microseconds instead of nanoseconds.
tarantool> interval.new({usec = -3000000})
---
- -2.147483648 seconds
...
-- Wrong sign via sec + nsec instead of just nsec like above.
tarantool> interval.new({sec=-1, nsec=1})
---
- +0.999999999 seconds
...
```
Although the first example is arguable. It just looks very counter-intuitive. We are not in Rome, we don't need to write -1 + 5 to get 4 (IV). I would rather expect that `-1.000000001` is expressed as -1 sec and -1 nsec. However maybe this crazy notation is some kind of standard and @Mons won't be against it. Hereby the great lord @Mons is being summoned.

Reply:
Well, these good counter-examples required _some_ changes in the way I display intervals, but now they seem to be displayed reasonably. First - how it behaves now:
```
tarantool> i1 = interval.new({nsec = -999999999})
---
...
tarantool> i2 = interval.new({nsec = -2})
---
...
tarantool> i = i1 + i2
---
...
tarantool> i1
---
- -0.999999999 seconds
...
tarantool> i2
---
- -0.000000002 seconds
...
tarantool> i.sec, i.nsec
---
- 0
- -1000000001
...
tarantool> interval.new({nsec = -3000000000})
---
- error: value -3000000000 of nsec is out of allowed range [-2147483647, 2147483647]
...
tarantool> interval.new({nsec = -300000000})
---
- -0.300000000 seconds
...
tarantool>  interval.new({usec = -3000000})
---
- error: value -3000000 of usec is out of allowed range [-2147483, 2147483]
...
tarantool> interval.new({sec=-1, nsec=1})
---
- -1.999999999 seconds
...

```

i.e.  -300000000 was way larger than 32-bit (int) nanoseconds could handle, so I bail out with error if overflow/underflow occurred while assigning to `nsec` field. (Not only via `nsec` itself, but also via `usec` or `msec`)

Reply:
Now to the code changes:


<details>
  <summary>interval display improvements</summary>

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -626,26 +626,51 @@ local function qtail(s)
     return #s ~= 0 and (s .. ', ') or s
 end
 
--- if sec is negative number, and nsec normalized then we denormalize sec
--- and move toward 0
-local function normalize_neg_nsec(sec, nsec)
-    if sec < 0 and nsec ~= 0 then
-        assert(nsec > 0)
-        nsec = NANOS_PER_SEC - nsec
-        sec = sec + 1
+-- if |nsec| is larger than allowed range 1e9 then modify passed
+-- sec accordingly
+local function denormalize_interval_nsec(sec, nsec)
+    if nsec == 0 then
+        return sec, nsec
+    end
+    -- nothing to change:
+    -- - if there is small nsec with 0 in sec
+    -- - or if both sec and nsec have the same sign, and nsec is
+    --   small enough
+    local same_sign = (sec < 0) == (nsec < 0)
+    if (sec == 0 or same_sign) and math_abs(nsec) < 1e9 then
+        return sec, nsec
+    end
+
+    sec = sec + math_modf(nsec / NANOS_PER_SEC)
+    if sec >= 0 then
+        nsec = nsec % NANOS_PER_SEC
+    else
+        nsec = NANOS_PER_SEC - nsec % NANOS_PER_SEC
     end
     return sec, nsec
 end
 
 -- signed or unsigned textual representation of sec and nsec
 local function seconds_str(need_sign, sec, nsec)
-    local prefix = need_sign and '%+d' or '%d'
-    sec, nsec = normalize_neg_nsec(sec, nsec)
     sec = math_fmod(sec, 60)
-    return nsec ~= 0 and (prefix..'.%09d seconds'):format(sec, nsec) or
-           (prefix..' seconds'):format(sec)
+    sec, nsec = denormalize_interval_nsec(sec, nsec)
+    local is_negative = sec < 0 or (sec == 0 and nsec < 0)
+    local str = is_negative and '-' or (need_sign and '+' or '')
+    sec = math_abs(sec)
+    if nsec ~= 0 then
+        str = str .. ('%d.%09d seconds'):format(sec, math_abs(nsec))
+    else
+        str = str .. ('%d seconds'):format(sec)
+    end
+    return str
 end
 
+
+local signed_fmt = {
+    [false] = '%d',
+    [true]  = '%+d',
+}
+
 --[[
     Convert to text interval values of different types
 
@@ -663,29 +688,31 @@ end
 local function interval_tostring(o)
     check_interval(o, 'datetime.interval.tostring')
     local s = ''
+    local need_sign = true
     if o.year ~= 0 then
         s = qtail(s) .. ('%+d years'):format(o.year)
+        need_sign = false
     end
     if o.month ~= 0 then
-        s = qtail(s) .. ('%+d months'):format(o.month)
+        s = qtail(s) .. (signed_fmt[need_sign] .. ' months'):format(o.month)
+        need_sign = false
     end
     local sec, nsec = o.sec, o.nsec
     if sec == 0 and nsec == 0 then
         return #s > 0 and s or '0 seconds'
     end
-    sec, nsec = normalize_nsec(sec, nsec)
     local abs_s = math_abs(sec)
     if abs_s < 60 then
-        s = qtail(s) .. seconds_str(true, sec, nsec)
+        s = qtail(s) .. seconds_str(need_sign, sec, nsec)
     elseif abs_s < (60 * 60) then
-        s = qtail(s) .. ('%+d minutes, %s'):
+        s = qtail(s) .. (signed_fmt[need_sign] .. ' minutes, %s'):
                         format(o.min, seconds_str(false, sec, nsec))
     elseif abs_s < SECS_PER_DAY then
-        s = qtail(s) .. ('%+d hours, %d minutes, %s'):
+        s = qtail(s) .. (signed_fmt[need_sign] .. ' hours, %d minutes, %s'):
                         format(o.hour, math_fmod(o.min, 60),
                                seconds_str(false, sec, nsec))
     else
-        s = qtail(s) .. ('%+d days, %d hours, %d minutes, %s'):
+        s = qtail(s) .. (signed_fmt[need_sign] .. ' days, %d hours, %d minutes, %s'):
                         format(o.day, math_fmod(o.hour, 24),
                                math_fmod(o.min, 60),
                                seconds_str(false, sec, nsec))
@@ -784,7 +811,8 @@ local function datetime_interval_sub(lhs, rhs)
         year, month, secs, nsec = interval_decouple_args(rhs)
         obj.year = obj.year - year
         obj.month = obj.month - month
-        obj.sec, obj.nsec = normalize_nsec(obj.sec - secs, obj.nsec - nsec)
+        -- do not normalize nsec yet - leave till the operation with date
+        obj.sec, obj.nsec = obj.sec - secs, obj.nsec - nsec
         return obj
     else
         error_incompatible("operator -")
@@ -821,7 +849,8 @@ local function datetime_interval_add(lhs, rhs)
         year, month, secs, nsec = interval_decouple_args(rhs)
         obj.year = obj.year + year
         obj.month = obj.month + month
-        obj.sec, obj.nsec = normalize_nsec(obj.sec + secs, obj.nsec + nsec)
+        -- do not normalize nsec yet - leave till the operation with date
+        obj.sec, obj.nsec = obj.sec + secs, obj.nsec + nsec
         return obj
     else
         error_incompatible("operator +")
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index f5fab702f..d28f0efcc 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -556,13 +556,13 @@ test:test("Time interval tostring()", function(test)
         {{week = 10, hour = 8, min = 10, sec = 30},
          '+70 days, 8 hours, 10 minutes, 30 seconds'},
         {{month = 20, week = 10, hour = 8, min = 10, sec = 30},
-         '+20 months, +70 days, 8 hours, 10 minutes, 30 seconds'},
+         '+20 months, 70 days, 8 hours, 10 minutes, 30 seconds'},
         {{year = 10, month = 20, week = 10, hour = 8, min = 10, sec = 30},
-         '+10 years, +20 months, +70 days, 8 hours, 10 minutes, 30 seconds'},
+         '+10 years, 20 months, 70 days, 8 hours, 10 minutes, 30 seconds'},
         {{year = 1e4, month = 20, week = 10, hour = 8, min = 10, sec = 30},
-         '+10000 years, +20 months, +70 days, 8 hours, 10 minutes, 30 seconds'},
+         '+10000 years, 20 months, 70 days, 8 hours, 10 minutes, 30 seconds'},
         {{year = 5e6, month = 20, week = 10, hour = 8, min = 10, sec = 30},
-         '+5000000 years, +20 months, +70 days, 8 hours, 10 minutes, 30 seconds'},
+         '+5000000 years, 20 months, 70 days, 8 hours, 10 minutes, 30 seconds'},
         {{year = -5e6, month = -20, week = -10, hour = -8, min = -10, sec = -30},
          '-5000000 years, -20 months, -70 days, -8 hours, -10 minutes, -30 seconds'},
         {{month = -20, week = -10, hour = -8, min = -10, sec = -30},
@@ -588,7 +588,7 @@ test:test("Time interval __index fields", function(test)
     local ival = date.interval.new{year = 12345, month = 123, week = 100,
                                    day = 45, hour = 48, min = 3, sec = 1,
                                    nsec = 12345678}
-    test:is(tostring(ival), '+12345 years, +123 months, +747 days, 0 hours,'..
+    test:is(tostring(ival), '+12345 years, 123 months, 747 days, 0 hours,'..
             ' 3 minutes, 1.012345678 seconds', '__tostring')
 
     test:is(ival.nsec, 12345678, 'nsec')
```  
</details>

<details>
  <summary>range checks for nsec</summary>

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -120,6 +120,9 @@ local MAX_DAY_RANGE = MAX_YEAR_RANGE * AVERAGE_DAYS_YEAR
 local MAX_HOUR_RANGE = MAX_DAY_RANGE * 24
 local MAX_MIN_RANGE = MAX_HOUR_RANGE * 60
 local MAX_SEC_RANGE = MAX_DAY_RANGE * SECS_PER_DAY
+local MAX_NSEC_RANGE = INT_MAX
+local MAX_USEC_RANGE = math_floor(MAX_NSEC_RANGE / 1e3)
+local MAX_MSEC_RANGE = math_floor(MAX_NSEC_RANGE / 1e6)
 local DEF_DT_ADJUST = builtin.DT_LIMIT
 
 local date_tostr_stash =
@@ -302,9 +305,9 @@ end
 
 -- check v value against maximum/minimum possible values
 -- if there is nil then simply return default 0
-local function checked_max_value(v, max, txt)
+local function checked_max_value(v, max, txt, def)
     if v == nil then
-        return 0
+        return def
     end
     if type(v) ~= 'number' then
         error(('numeric value expected, but received %s'):
@@ -339,22 +342,24 @@ local function interval_decouple_args(obj)
     if is_interval(obj) then
         return obj.year, obj.month, obj.sec, obj.nsec, obj.adjust
     end
-    local year = checked_max_value(obj.year, MAX_YEAR_RANGE, 'year')
-    local month = checked_max_value(obj.month, MAX_MONTH_RANGE, 'month')
+    local year = checked_max_value(obj.year, MAX_YEAR_RANGE, 'year', 0)
+    local month = checked_max_value(obj.month, MAX_MONTH_RANGE, 'month', 0)
     local adjust = adjust_xlat[obj.adjust] or DEF_DT_ADJUST
 
     local secs = 0
     secs = secs + (7 * SECS_PER_DAY) *
-                  checked_max_value(obj.week, MAX_WEEK_RANGE, 'week')
-    secs = secs + SECS_PER_DAY * checked_max_value(obj.day, MAX_DAY_RANGE, 'day')
-    secs = secs + (60 * 60) * checked_max_value(obj.hour, MAX_HOUR_RANGE, 'hour')
-    secs = secs + 60 * checked_max_value(obj.min, MAX_MIN_RANGE, 'min')
-    local sec = checked_max_value(obj.sec, MAX_SEC_RANGE, 'sec')
+                  checked_max_value(obj.week, MAX_WEEK_RANGE, 'week', 0)
+    secs = secs + SECS_PER_DAY *
+                  checked_max_value(obj.day, MAX_DAY_RANGE, 'day', 0)
+    secs = secs + (60 * 60) *
+                  checked_max_value(obj.hour, MAX_HOUR_RANGE, 'hour', 0)
+    secs = secs + 60 * checked_max_value(obj.min, MAX_MIN_RANGE, 'min', 0)
+    local sec = checked_max_value(obj.sec, MAX_SEC_RANGE, 'sec', 0)
     check_integer(sec, 'sec')
-    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
-    check_integer(nsec, 'nsec')
-    check_integer(usec, 'usec')
-    check_integer(msec, 'msec')
+
+    local nsec = checked_max_value(obj.nsec, MAX_NSEC_RANGE, 'nsec')
+    local usec = checked_max_value(obj.usec, MAX_USEC_RANGE, 'usec')
+    local msec = checked_max_value(obj.msec, MAX_MSEC_RANGE, 'msec')
     local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
                        bool2int(msec ~= nil)
     if count_usec > 1 then
```  
</details>